crates/bevy_ecs/src/bundle/info.rs

```diff
@@ -543,12 +624,96 @@ impl Bundles {
             });
         *bundle_id
     }
+
+    /// Checks if the bundles containing `requiree` can have their required components be updated,
+    /// in which case this returns `Ok(true)`.
+    ///
+    /// Returns `Ok(false)` if there are no known bundles with this component.
+    ///
+    /// Returns the [`RequiredComponentsError::RemovedFromArchetype`] error if bundles cannot be updated
+    /// in which case the registration of any new required component must be refused.
+    pub(crate) fn verify_to_refresh_required_components(
+        &self,
+        requiree: ComponentId,
+    ) -> Result<bool, RequiredComponentsError> {
+        let Some(bundles_with_requiree) = self
+            .components_in_bundles
+            .get(requiree.index())
+            .filter(|bundles| !bundles.is_empty())
+        else {
+            // no bundle with `requiree` exists
+            return Ok(false);
+        };
+
+        // `EntityWorldMut::remove_with_requires` generate edges between archetypes where the required
```

Comment:
This invariant feels very hidden, I would add a comment on `EntityWorldMut::remove_with_requires` explaining that this part of the code relies on it populating bundles in such a way that they are inserted in `contributed_bundle_ids`.